realm/src/main/java/io/realm/Realm.java

```diff
@@ -428,6 +437,15 @@ private static Realm createAndValidate(String absolutePath, byte[] key, boolean
         if (references == null) {
             references = 0;
         }
+        if (references == 0) {
+            AtomicInteger counter = openRealms.get(id);
```

Comment:
This doesn't look threadsafe, two threads can call this before the value is inserted, making the counter miss one. Testing for null and inserting new AtomicInteger must be a atomic operation. I don't think there is any way around a synchronized block for this?